SCons/CacheDir.py

```diff
@@ -50,18 +51,28 @@ def CacheRetrieveFunc(target, source, env):
         cd.CacheDebug('CacheRetrieve(%s):  %s not in cache\n', t, cachefile)
         return 1
     cd.hits += 1
-    cd.CacheDebug('CacheRetrieve(%s):  retrieving from %s\n', t, cachefile)
     if SCons.Action.execute_actions:
         if fs.islink(cachefile):
             fs.symlink(fs.readlink(cachefile), t.get_internal_path())
         else:
-            env.copy_from_cache(cachefile, t.get_internal_path())
+            try:
+                env.copy_from_cache(cachefile, t.get_internal_path())
+            except (shutil.SameFileError, IOError) as e:
```

Comment:
I don't think the `SameFileError` is going to leave a remnant, as it's raised before any copying is attempted.